### PR TITLE
Implement H5Type for complex

### LINKED ISSUE #208: Store num_complex datatypes
Hi,

I am wondering what the best way is to store a `Complex<f32>` in a hdf5 dataset.

Since the type is defined in the num_complex crate, I cannot easily derive H5Type. Do I need to define a wrapper type? Or, can this be added to hdf5-rust maybe as a Cargo feature?

What other options do I have?

### PR DESCRIPTION
**Pull request by mulimoen**
Closes #208

### CODE REVIEW

File: hdf5-types/Cargo.toml [hunk 1]
```diff
@@ -14,12 +14,14 @@ categories = ["encoding"]
...
 [features]
 h5-alloc = []
+complex = ["num-complex"]
 
 [dependencies]
 ascii = "1.0"
 libc = "0.2"
 hdf5-sys = { version = "0.8.1", path = "../hdf5-sys" }  # !V
 cfg-if = "1.0.0"
+num-complex = { version = "0.4.0", optional = true, default-features = false }
```
Comment (aldanor) [inline hdf5-types/Cargo.toml:24]: @mulimoen Maybe set to `"0.4"`? E.g. because 0.4.2 is already out.

File: hdf5-types/src/complex.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+use super::{CompoundField, CompoundType, H5Type, TypeDescriptor};
+
+use std::mem::size_of;
+
+use num_complex::Complex;
+
+unsafe impl<T: H5Type> H5Type for Complex<T> {
+    fn type_descriptor() -> TypeDescriptor {
+        TypeDescriptor::Compound(CompoundType {
```
Comment (aldanor) [inline hdf5-types/src/complex.rs:9]: This is inherently unsafe, how do we know that offsets are right? (e.g. should we check actual offsets here?)

Comment (mulimoen) [reply hdf5-types/src/complex.rs:9]: I am not sure how we would check the offsets

File: hdf5-types/Cargo.toml (hunk 1 above)
Comment (mulimoen) [reply hdf5-types/Cargo.toml]: Set to 0.4

File: hdf5-types/src/complex.rs (hunk 2 above)
Comment (mulimoen) [reply hdf5-types/src/complex.rs:9]: That would require us to make something to point at. We could make an instance of `MaybeUninit`, but field access is only a future feature

Comment (aldanor) [reply hdf5-types/src/complex.rs:9]: Hm... I guess we could do something like this, does that make sense? https://rust.godbolt.org/z/TWbavj98G

Comment (mulimoen) [reply hdf5-types/src/complex.rs:9]: We could do
```rust
unsafe impl<T: H5Type> H5Type for Complex<T> {
    fn type_descriptor() -> TypeDescriptor {
        use std::mem::MaybeUninit;
        // MaybeUninit has the same size and is transparent, but we can
        // do some pointer arithmetics on a real instance
        let dummystruct: Complex<MaybeUninit<T>> =
            Complex::new(MaybeUninit::uninit(), MaybeUninit::uninit());

        assert_eq!(size_of_val(&dummystruct), size_of::<Complex<T>>());

        let base = addr_of!(dummystruct).cast::<u8>();
        let ptr_r = addr_of!(dummystruct.re).cast::<u8>();
        let ptr_i = addr_of!(dummystruct.im).cast::<u8>();

        TypeDescriptor::Compound(CompoundType {
            fields: vec![
                // Compatible with h5py definition of complex
                CompoundField::typed::<T>("r", unsafe { ptr_r.offset_from(base) } as usize, 0),
                CompoundField::typed::<T>("i", unsafe { ptr_i.offset_from(base) } as usize, 1),
            ],
            size: size_of::<Complex<T>>(),
        })
    }
}
```

Comment (mulimoen) [reply hdf5-types/src/complex.rs:9]: Might be enough to link to the documentation on representation? https://docs.rs/num-complex/0.4.2/num_complex/struct.Complex.html#representation-and-foreign-function-interface-compatibility

Comment (aldanor) [reply hdf5-types/src/complex.rs:9]: Maybe add this link in a comment in code? E.g. "`Complex<T>` should be FFI-equivalent to `[T; 2]`: <link>".

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
 - Support field renaming via `#[hdf5(rename = "new_name")]` helper attribute.
 - Add a `ByteReader` which implements `std::io::{Read, Seek}` for 1D `u8`
   datasets. Usage via `Dataset::as_byte_reader()`.
+- `H5Type` for `Complex` struct from `num_complex`
```
Comment (aldanor) [inline CHANGELOG.md:12]:
```suggestion
- Implement `H5Type` for `num_complex::Complex`.
```

**Comment by aldanor** [comment]
Minor nits aside, needs to be rebased on master (because of the changelog).

File: hdf5-types/src/complex.rs (hunk 2 above)
Comment (mulimoen) [reply hdf5-types/src/complex.rs:11]: Thanks, added this as a comment under `type_descriptor()`

**Review by aldanor — APPROVED** [review]

**Pull request merged**